ZEngine/src/Helper.cpp

```diff
@@ -0,0 +1,879 @@
+        specification.ColorBlendAttachmentState.dstAlphaBlendFactor = VK_BLEND_FACTOR_ONE_MINUS_SRC_ALPHA;
+        specification.ColorBlendAttachmentState.alphaBlendOp        = VK_BLEND_OP_ADD;
+
+        // this configuration is for global color blending settings
+        specification.ColorBlendStateCreateInfo.sType             = VK_STRUCTURE_TYPE_PIPELINE_COLOR_BLEND_STATE_CREATE_INFO;
+        specification.ColorBlendStateCreateInfo.logicOpEnable     = VK_FALSE;
+        specification.ColorBlendStateCreateInfo.logicOp           = VK_LOGIC_OP_COPY; // Optional
+        specification.ColorBlendStateCreateInfo.attachmentCount   = 1;
+        specification.ColorBlendStateCreateInfo.pAttachments      = &specification.ColorBlendAttachmentState;
+        specification.ColorBlendStateCreateInfo.blendConstants[0] = 0.0f; // Optional
+        specification.ColorBlendStateCreateInfo.blendConstants[1] = 0.0f; // Optional
+        specification.ColorBlendStateCreateInfo.blendConstants[2] = 0.0f; // Optional
+        specification.ColorBlendStateCreateInfo.blendConstants[3] = 0.0f; // Optional
+        specification.ColorBlendStateCreateInfo.flags             = 0;
+        specification.ColorBlendStateCreateInfo.pNext             = nullptr;
+
+        /*Pipeline layout*/
+        // Uniform Shader configuration
+        specification.LayoutCreateInfo.sType                  = VK_STRUCTURE_TYPE_PIPELINE_LAYOUT_CREATE_INFO;
+        specification.LayoutCreateInfo.setLayoutCount         = 0;       // Optional
+        specification.LayoutCreateInfo.pSetLayouts            = nullptr; // Optional
+        specification.LayoutCreateInfo.pushConstantRangeCount = 0;       // Optional
+        specification.LayoutCreateInfo.pPushConstantRanges    = nullptr; // Optional
+        specification.LayoutCreateInfo.flags                  = VK_PIPELINE_LAYOUT_CREATE_INDEPENDENT_SETS_BIT_EXT;
+        specification.LayoutCreateInfo.pNext                  = nullptr;
+
+        /*Vertex Input*/
+        specification.VertexInputStateCreateInfo.sType                           = VK_STRUCTURE_TYPE_PIPELINE_VERTEX_INPUT_STATE_CREATE_INFO;
+        specification.VertexInputStateCreateInfo.vertexAttributeDescriptionCount = 0;
+        specification.VertexInputStateCreateInfo.pVertexAttributeDescriptions    = nullptr;
+        specification.VertexInputStateCreateInfo.vertexBindingDescriptionCount   = 0;
+        specification.VertexInputStateCreateInfo.pVertexBindingDescriptions      = nullptr;
+        specification.VertexInputStateCreateInfo.flags                           = 0;
+        specification.VertexInputStateCreateInfo.pNext                           = nullptr;
+    }
+
+    void TransitionImageLayout(Hardwares::VulkanDevice& device, VkImage image, VkFormat image_format, VkImageLayout old_image_layout, VkImageLayout new_image_layout)
+    {
+        auto                 graphic_queue  = device.GetAnyGraphicQueue();
```

Comment:
There are multiple issues I'm seeing with TransitionImageLayout:

1. You shouldn't submit command buffers containing only single transitions. Rather, layout transitions are supposed to occur between different sets of commands in the same command buffer, or different command buffers on the same queue, or different command buffers on different queues. While it is not wrong to use them like this it is also wasteful to submit a single command buffer for each transition. Try including the transitions in your other command buffers, such as the transfer command buffer.
2. The way I understand GetAnyGraphicQueue is that each time a different queue can be returned. The only reason this works right now is because you are lockstepping the submits. Imagine the function being used to transition the image layout multiple times, but not waiting for the queues to be idle. Which transition would be executed first? In which layout would the image then be?

Reply:
Removed those api.
thanks!